server/matchmaker/matchmaker_queue.py

```diff
@@ -19,16 +19,70 @@ def __init__(
         self.queue_name = queue_name
         self.rating_prop = 'ladder_rating'
         self.queue = OrderedDict()
+        self._matches = deque()
+        self._shutdown = False
         self._logger.debug("MatchmakerQueue initialized for %s", queue_name)
 
-    def push(self, search: Search):
+    async def iter_matches(self):
+        """ Asynchronusly yields matches as they become available """
+
+        while not self._shutdown:
+            if not self._matches:
+                await asyncio.sleep(1)
+                continue
+
+            yield self._matches.popleft()
+
+    async def search(self, search: Search):
         """
-        Push the given search object onto the queue
+        Search for a match.
 
-        :param search:
-        :return:
+        If a suitable match is found, immediately calls on_matched_with on both players.
+
+        Otherwise, puts a search object into the Queue and awaits completion
+
+        :param player: Player to search for a matchup for
         """
-        self.queue[search.player] = search
+        assert search is not None
+
+        with server.stats.timer('matchmaker.search'):
+            try:
+                if self.find_match(search):
+                    return
+
+                self._logger.debug("Found nobody searching, pushing to queue: %s", search)
+                self.push(search)
+                await search.await_match()
+                self._logger.debug("Search complete: %s", search)
+            except CancelledError:
+                pass
+            finally:
+                # If the queue was cancelled, or some other error occured,
+                # make sure to clean up.
+                self.game_service.mark_dirty(self)
+                if search in self.queue:
+                    del self.queue[search]
+
+    def find_match(self, search: Search) -> bool:
+        self._logger.debug("Searching for matchup for %s", search.players)
+        for other in self.queue.copy().values():
+            if other == search:
```

Comment:
Yes it works. This is old code again, it shows up as new lines in git because I refactored it to a function, but it is exactly the same code that is running on prod right now

Reply:
`copy` always does a shallow copy. For deep copy you have to use `deepcopy`